### DISCUSSION TITLE
what time can support windows?

### DISCUSSION THREAD

**Discussion by NiZerin** [lima-vm/lima#391] [Q&A] [answered]
just like title, what time can support windows?

**Comment by AkihiroSuda** [comment, marked as answer]
Not in foreseeable future, but depends on contribution from the community

**Reply by route33** [reply]
What about under MSYS2?

❯ make && make install
mkdir -p _output/bin
cp -a ./cmd/lima _output/bin/lima
# The hostagent must be compiled with CGO_ENABLED=1 so that net.LookupIP() in the DNS server
# calls the native resolver library and not the simplistic version in the Go library.
CGO_ENABLED=1 go build -ldflags="-s -w -X github.com/lima-vm/lima/pkg/version.Version=v0.7.4-4-g61d57bb" -o _output/bin/limactl ./cmd/limactl
# github.com/lima-vm/lima/pkg/lockutil
pkg\lockutil\lockutil.go:34:27: undefined: unix.LOCK_EX
pkg\lockutil\lockutil.go:38:28: undefined: unix.LOCK_UN
pkg\lockutil\lockutil.go:48:10: undefined: unix.Flock
pkg\lockutil\lockutil.go:49:27: undefined: unix.EINTR
# github.com/lima-vm/lima/pkg/networks
pkg\networks\validate.go:76:25: undefined: syscall.Stat_t
make: *** [Makefile:44: _output/bin/limactl] Error 2

**Reply by afbjorklund** [reply]
Msys2 is Windows, but you might try WSL instead ? (it is Linux and similar to Lima)

**Reply by route33** [reply]
Yeah, I could tell the story of my life, I'm a mac user, forced by the company I work for to use Windows. WSL is not so great and got stuck with msys2 for the bare minimum.
I'm going to try the old fashioned VirtualBox.
Thanks anyway.

**Reply by afbjorklund** [reply]
QEMU does work on Windows, I have a basic VM running - but not all the bells and whistles of Lima...
If you want to play along, there are QEMU binaries with WHPX available at: https://qemu.weilnetz.de/

Initially one effort will be to change the "unix socket" over to a "named pipe", for running on Windows.
I like msys2 and brew, they give a more familiar environment when working in Windows and macOS.

**Comment by jandubois** [comment]
@route33 Depending on your needs you could check out [Rancher Desktop](https://rancherdesktop.io) on Windows. It uses Lima on Linux and macOS, and WSL2 on Windows to provide a container runtime and k8s.

It is also an open source project, and but still relatively newish, so may not yet provide exactly what you want.

**Reply by afbjorklund** [reply]
It would also be interesting to know what makes WSL2 "not so great", and how Lima makes (or could make, perhaps) it better ?

**Reply by afbjorklund** [reply]
My main Windows (or Mac) experience is from docker-machine and podman-machine, when using the qemu (or virtualbox) driver

**Reply by jandubois** [reply]
I don't have a comprehensive list, just some observations:

* All WSL2 "distros" run inside namespaces inside the same VM. They share CPU and memory limits and the network (not namespaced). This is both good and bad. 😄 
* The IP address of your VM changes every time you restart Windows (or maybe even just the VM), which can be a pain for certs.
* There are various unresolved conflicts with VPNs that you have to work around. Not sure who is to blame here (which may be the reason it is still not fixed).
* WSL2 doesn't have an init system
* Each distro is put to sleep quite aggressively if WSL2 determines there is nothing running in it.

There is probably more; this is from the top of my head. I still think it is pretty great, and really good enough that I don't see a reason why you would need Lima on Windows.

One big advantage for WSL2 is that it can use Hyper-V even with Windows Home. To run your own VMs on Hyper-V you would otherwise need at least a Windows Pro license.

**Reply by afbjorklund** [reply]
I guess it (WSL model) is more similar in usage to Docker Desktop, as opposed to the "machine" model from Docker Toolbox

I have only casually compared it to Cygwin, so not the best judge... But I did notice that the container-based WSL1 "failed".

**Reply by afbjorklund** [reply]
My main reason for trying to run Podman (and maybe eventually also Lima?) on Windows, was more of a "why not" ... 🙃

It didn't have a use case, and most certainly not a business case. All I know so far is that QEMU works. And ssh, I suppose.

**Reply by jandubois** [reply]
> All I know so far is that QEMU works. And ssh, I suppose.

Yeah, but what are the advantages over using just WSL2, which has already "solved" the issue of shared filesystems for you? And doesn't really need `ssh` because you can already run `wsl -d MYDISTRO uname -a` or whatever.

**Reply by afbjorklund** [reply]
I don't really know how WSL2 and `wsl` are being implemented, but I do know where to look and what syntax to use for lima and friends... In terms of filesystem I think it didn't even have virtfs, so most likely it would be sshfs (at best). I know slirp (user net) works, not sure about the rest. And Red Hat came to a similar conclusion about Podman, they use WSL2 ([*](https://www.redhat.com/sysadmin/podman-windows-wsl2)) and don't support Windows.
